### PR TITLE
Annotations for state_deltas.py

### PR DESCRIPTION
**Pull request by DMRobertson**
I was sad that I couldn't do better for
`_curr_state_delta_stream_cache`. At least it's explicitly called out in
a comment with #TODO.

### CODE REVIEW

File: mypy.ini [hunk 1]
```diff
@@ -183,6 +182,9 @@ disallow_untyped_defs = True
 [mypy-synapse.storage.databases.main.room_batch]
 disallow_untyped_defs = True
 
+[mypy-synapse.storage.databases.main.state_deltas]
+disallow_untyped_defs = True
```
Comment (clokep) [inline mypy.ini:186]: See #11322 instead. 😄

File: synapse/storage/databases/main/state_deltas.py [hunk 2]
```diff
@@ -16,11 +16,17 @@
...
+from synapse.storage.database import LoggingTransaction
+from synapse.util.caches.stream_change_cache import StreamChangeCache
 
 logger = logging.getLogger(__name__)
 
 
 class StateDeltasStore(SQLBaseStore):
+    # This class must be mixed in with a child class which provides the following
+    # attribute. TODO: can we get static analysis to enforce this?
+    _curr_state_delta_stream_cache: StreamChangeCache
```
Comment (reivilibre) [inline synapse/storage/databases/main/state_deltas.py:28]: Can we make this an abstract class perhaps, with this being an abstract.. field... if that exists ...

It doesn't seem like it does, but I would be tempted to try an abstract property:

```python
    @property
    @abstractmethod
    def _curr_state_delta_stream_cache(self) -> StreamChangeCache:
        ...
```

Not super clean, but curious if that'd do the trick (and if it's allowed to just overwrite it with a field later)

**Review by reivilibre — APPROVED** [review]
Hacky abstract thing is up to you, everything else LGTM (though the mypy.ini change is soon going to be obsolete one way or the other ...)

File: synapse/storage/databases/main/state_deltas.py (hunk 2 above)
Comment (DMRobertson) [reply synapse/storage/databases/main/state_deltas.py:28]: I got

```
synapse/app/generic_worker.py:222: error: Definition of "_curr_state_delta_stream_cache" in base class "StateDeltasStore" is incompatible with definition in base class "SlavedEventStore"  [misc]
```

after

```
class StateDeltasStore(SQLBaseStore, abc.ABC):
    @property
    @abc.abstractmethod
    def _curr_state_delta_stream_cache(self) -> StreamChangeCache:
        """This class must be mixed in with a child class which provides the
        following attribute."""
        ...
```

Comment (DMRobertson) [reply synapse/storage/databases/main/state_deltas.py:28]: You might be interested in the discussion here: https://github.com/python/mypy/issues/4426

**Pull request merged**